webrender/src/frame.rs

```diff
@@ -49,14 +50,52 @@ impl<'a> FlattenContext<'a> {
             builder: builder,
             resource_cache: resource_cache,
             replacements: Vec::new(),
+            nested_display_list_info: Vec::new(),
         }
     }
 
-    fn clip_id_with_replacement(&self, id: ClipId) -> ClipId {
-        match self.replacements.last() {
+    fn push_nested_display_list_ids(&mut self, info: ClipAndScrollInfo) {
+        self.nested_display_list_info.push((info.scroll_node_id, info.clip_node_id()));
+    }
+
+    fn pop_nested_display_list_ids(&mut self) {
+        self.nested_display_list_info.pop();
+    }
+
+    fn clip_and_scroll_info_with_replacements(&self, info: &mut ClipAndScrollInfo) {
+        info.scroll_node_id = self.scroll_frame_id_with_replacement(info.scroll_node_id);
+
+        let mut clip_node_id = match info.clip_node_id {
+            Some(id) => id,
+            None => return,
+        };
+
+        for nested_list_info in self.nested_display_list_info.iter().rev() {
+            if clip_node_id.is_root_scroll_node() {
+                clip_node_id = nested_list_info.1;
+            } else {
+                break;
+            }
+        }
```

Comment:
Just checking my understanding here: it should be possible to exit this loop with `clip_node_id.is_root_scroll_node() == true`, right? That would be if there weren't any nested scrolling contexts at all?